src/scanForInputs.js

```diff
@@ -1,42 +1,55 @@
-const Form = require('./Form')
+const Form = require('./Form/Form')
 const {notifyWebApp} = require('./autofill-utils')
+const {EMAIL_SELECTOR, PASSWORD_SELECTOR, FIELD_SELECTOR, SUBMIT_BUTTON_SELECTOR} = require('./Form/selectors')
+
+const forms = new Map()
 
 // Accepts the DeviceInterface as an explicit dependency
 const scanForInputs = (DeviceInterface) => {
-    const forms = new Map()
-
-    const EMAIL_SELECTOR = `
-            input:not([type])[name*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input[type=""][name*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input[type=text][name*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input:not([type])[id*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input:not([type])[placeholder*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input[type=""][id*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input[type=text][placeholder*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input[type=""][placeholder*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input:not([type])[placeholder*=mail i]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input[type=email]:not([readonly]):not([disabled]):not([hidden]):not([aria-hidden=true]),
-            input[type=text][aria-label*=mail i],
-            input:not([type])[aria-label*=mail i],
-            input[type=text][placeholder*=mail i]:not([readonly])
-        `
-
-    const addInput = input => {
-        const parentForm = input.form
+    const getParentForm = (input) => {
+        if (input.form) return input.form
+
+        let element = input
+        // traverse the DOM to search for related inputs
+        while (element !== document.body) {
+            element = element.parentElement
+            const inputs = element.querySelectorAll(FIELD_SELECTOR)
+            const buttons = element.querySelectorAll(SUBMIT_BUTTON_SELECTOR)
+            // If we find a button or another input, we assume that's our form
+            if (inputs.length > 1 || buttons.length) {
```

Comment:
❓ Just confirming, there needs to be more than one input for this to be detected as a form?

Reply:
We start traversing the DOM from the `input` up, so at least one `input` is always present. We might need a smarter version of this functionality, but this covers the vast majority of cases, so I'm happy with the low-hanging-fruit at this stage.